File: dashboard/android/app/src/main/java/com/attobot/dashboard/ui/components/DataTable.kt
package com.attobot.dashboard.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.attobot.dashboard.ui.theme.border
import com.attobot.dashboard.ui.theme.muted
import com.attobot.dashboard.ui.theme.panel2

/** A table column: header label, fixed [width], and the cell renderer. */
data class TableColumn<T>(
    val header: String,
    val width: Dp,
    val cell: @Composable RowScope.(T) -> Unit,
)

/**
 * Wide tables scroll horizontally as a block, exactly like the web/RN tables.
 * Rows render as a flat list (admin views are small enough that lazy
 * virtualisation isn't worth the nested-scroll complexity). Rows are tappable
 * when [onRow] is supplied.
 */
@Composable
fun <T> DataTable(
    columns: List<TableColumn<T>>,
    rows: List<T>,
    onRow: ((T) -> Unit)? = null,
    emptyText: String = "No rows.",
) {
    val tableWidth = columns.sumOf { it.width.value.toDouble() }.dp
    val scroll = rememberScrollState()
    Column(
        Modifier
            .fillMaxWidth()
            .horizontalScroll(scroll)
            .width(tableWidth),
    ) {
        // Header row (panel2 background).
        Row(Modifier.fillMaxWidth().background(panel2)) {
            columns.forEach { col ->
                Box(
                    Modifier
                        .width(col.width)
                        .padding(horizontal = 10.dp, vertical = 8.dp),
                ) {
                    Text(
                        col.header,
                        color = muted,
                        fontSize = 12.sp,
                        fontWeight = FontWeight.Medium,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
            }
        }
        HorizontalDivider(color = border)
        if (rows.isEmpty()) {
            Text(
                emptyText,
                color = muted,
                fontSize = 13.sp,
                textAlign = TextAlign.Center,
                modifier = Modifier.fillMaxWidth().padding(vertical = 24.dp),
            )
        } else {
            rows.forEachIndexed { i, row ->
                Row(
                    Modifier
                        .fillMaxWidth()
                        .clickable(enabled = onRow != null) { onRow?.invoke(row) },
                ) {
                    val rowScope = this
                    columns.forEach { col ->
                        Box(
                            Modifier
                                .width(col.width)
                                .padding(horizontal = 10.dp, vertical = 8.dp),
                            contentAlignment = Alignment.CenterStart,
                        ) {
                            with(rowScope) { col.cell(row) }
                        }
                    }
                }
                if (i < rows.lastIndex) {
                    HorizontalDivider(color = border)
                }
            }
        }
    }
}

File: dashboard/ios/OpenWorldDashboard/Views/Components/DataTable.swift
import SwiftUI

/// Wide, horizontally-scrolling table. Rows are flat (admin views are small).
/// Mirrors the RN `DataTable`.
struct DataTableColumn<Row> {
    let id: String
    let title: String
    let width: CGFloat?
    let cell: (Row) -> AnyView

    init<C: View>(
        id: String,
        title: String,
        width: CGFloat? = nil,
        @ViewBuilder cell: @escaping (Row) -> C
    ) {
        self.id = id
        self.title = title
        self.width = width
        self.cell = { row in AnyView(cell(row)) }
    }
}

struct DataTable<Row: Hashable>: View {
    let columns: [DataTableColumn<Row>]
    let rows: [Row]
    var onRow: ((Row) -> Void)? = nil
    var emptyText: String = "No rows."

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            VStack(alignment: .leading, spacing: 0) {
                // Header
                HStack(spacing: 0) {
                    ForEach(columns, id: \.id) { col in
                        Text(col.title)
                            .font(.system(size: 12, weight: .medium))
                            .foregroundStyle(Theme.muted)
                            .lineLimit(1)
                            .frame(width: col.width, alignment: .leading)
                            .padding(.horizontal, 10)
                            .padding(.vertical, 8)
                    }
                }
                .background(Theme.panel2)
                .overlay(alignment: .bottom) {
                    Rectangle().fill(Theme.border).frame(height: 1)
                }
                if rows.isEmpty {
                    Text(emptyText)
                        .font(.system(size: 13))
                        .foregroundStyle(Theme.muted)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, 24)
                } else {
                    ForEach(Array(rows.enumerated()), id: \.offset) { _, row in
                        rowView(row)
                    }
                }
            }
        }
    }

    @ViewBuilder
    private func rowView(_ row: Row) -> some View {
        Group {
            if onRow != nil {
                Button {
                    onRow?(row)
                } label: {
                    rowContent(row)
                }
                .buttonStyle(RowHighlightStyle())
            } else {
                rowContent(row)
            }
        }
        .overlay(alignment: .bottom) {
            Rectangle().fill(Theme.border).frame(height: 1)
        }
    }

    private func rowContent(_ row: Row) -> some View {
        HStack(spacing: 0) {
            ForEach(columns, id: \.id) { col in
                col.cell(row)
                    .frame(width: col.width, alignment: .leading)
                    .padding(.horizontal, 10)
                    .padding(.vertical, 8)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }
}

/// Press-state highlight (panel2 on press) matching the RN `pressed` style.
struct RowHighlightStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .background(configuration.isPressed ? Theme.panel2 : Color.clear)
    }
}
